json-everything.net/wwwroot/md/vocabs-data-2022.md

```diff
@@ -0,0 +1,123 @@
+
+The value of `data` must be an object.  The keys of the object correspond to valid JSON Schema keywords, and the values MUST be one of
+
+- JSON Pointers per [RFC 6901](https://www.rfc-editor.org/rfc/rfc6901)
+- [Relative JSON Pointers](https://json-schema.org/draft/2019-09/relative-json-pointer.html)
+- URI references per [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986), potentially with JSON Pointer fragments
+
+The set of keywords in the schema or subschema containing `data` SHOULD be distinct from the set of keys defined within `data`.  Behavior in the event of an overlap is not defined.
+
+`data` operates in two phases:
+
+1. All of the values are dereferenced per sections 3.2 and 3.3.
+2. The keywords and values then form a schema which is applied to the instance at the current location.
+
+The validation and annotation results of `data` are those of the formed schema.  More detail regarding output can be found in section 3.4.
+
+### 3.2 Contextual Behavior
+
+`data` MUST be processed contextually in accordance with the draft of the schema in which it is used.  For example, if `data` is used in a schema that declares draft 2019-09, then
+
+- the keys SHOULD be recognized by draft 2019-09 or defined in other referenced vocabularies, and
+- the URI references MUST be resolved following the same rules as `$ref` as defined by draft 2019-09.
+
+If a key is not a recognized keyword in the schema's draft (e.g. `$dynamicAnchor` in a draft 2019-09 schema), then this key is ignored just as it would be ignored if it were in the schema or subschema containing `data`.
+
+### 3.3 URI Resolution
+
+The values of `data` are dereferenced in different ways depending on the format of the value.
+
+If the value is a JSON Pointer, it is resolved against the instance root.
+
+If the value is a Relative JSON Pointer, it is resolved against the instance at the location currently being evaluated.
+
+If the value is a URI reference, it must be resolved in accordance with the rules of `$ref` resolution for the containing spec, as per section 3.2.  The primary different between resolving URIs for `data` and resolving URIs for `$ref` is that `$ref` requires that the indicated data must represent a valid schema object whereas `data` can reference any valid JSON value.
+
+External document data can be referenced using an absolute URI base identifier along with a JSON Pointer fragment.
+
+Implementations SHOULD be able to download external JSON documents, but MAY provide means to pre-load and cache such documents against the identifiers by which they will be referenced.  Documents downloaded from URIs which contain a JSON Pointer fragment MUST be interpreted using a media type that allows resolution of such fragments.
+
+If any single reference cannot be resolved, validation MUST fail; otherwise the full value at the specified location is returned.
+
+If the resolved value is not valid for the associated keyword, validation MUST fail.
```

Comment:
> Which could be from an unknown vocabulary if this is combined with other extension vocabularies.

I may need to add a restriction that core-vocab keywords (`$*`) are explicitly disallowed.  I'm not sure what use having `$id` or `$schema` pulled from an external location would be.

Adding this restriction would have `data` use the parent schema's meta-schema and would implicitly require that any keywords are from the vocabs already in use.

Reply:
Ah, yeah, that is probably a good idea.  You don't want the template subschema to be something you can reference by URI, as it's not meaningful: only a specific resolved version would be a meaningful reference target.

So `$id` and `$schema` should be out.  `$ref` and `$dynamicRef` are problematic because a URI-reference value would be ambiguous (is it the `$ref` target or a `data` template location that provides the `$ref` target?), and `$ref`-ing some runtime data is like `eval()`-ing a random string of JavaScript code.  Definitely a security concern.